### PR TITLE
#160370617 Request a one way trip

### PR DESCRIPTION
**Pull request by Spectrumsun**
#### What does this PR do?
- A requester can make a travel request for a one-way trip

#### Description of Task to be completed?
- Given that the requester is logged in
- The requester should be able to save a new request
- Without the arrival date if the trip type is 'one way`

#### How should this be manually tested?
- Clone the repo -git clone https://github.com/andela/travel_tool_back.git
- Enter into the project directory - cd travel_tool_back
- Install project dependencies using `yarn install`
- Run database migrations using `yarn db:migrate`
- Seed the database using `yarn db:seed`
- Start the app using make start
- Login on your browser and grab your token from the application tab in Developers tab
- Open Postman and place your token on the authorization header
- Send a post request to '/api/v1/requests ' with the following fields
- `name`, `origin`, `manager`, `gender`, `department`, `role`, `departureDate`, `arrivalDate`, `destination `, `tripType`  

#### Any background context you want to provide?
- You have to specify the trip type when adding a new request with the field `tripType` and value `oneWay` for a one way trip
- We have three types of Trips, `return`, `oneWay`, `multi` 
- You have to specify the arrival date  if the trip type is `return` or  `multi`
#### What are the relevant pivotal tracker stories?
[#160370617](https://www.pivotaltracker.com/n/projects/2184887)

#### Screenshots (if appropriate)
image

#### Questions:
- None

### CODE REVIEW

File: src/middlewares/Validator.js [hunk 1]
```diff
@@ -81,4 +85,23 @@ export default class Validator {
     const errors = req.validationErrors();
     Validator.errorHandler(res, errors, next);
   }
+
+  static trip(req, res, next) {
+    const { arrivalDate } = req.body;
+    const oneWayTrip = (/true/i).test(req.body.oneWayTrip);
```
Comment (johnmutuma5) [inline src/middlewares/Validator.js]: @Spectrumsun, clever use of the regex test here. Nice.

File: src/modules/requests/__tests__/requests.test.js [hunk 2]
```diff
@@ -417,6 +413,85 @@ describe('Requests Controller', () => {
...
+        const res = await request(app)
+          .post('/api/v1/requests')
+          .set('authorization', token)
+          .send({ ...newRequest });
+        expect(res.status).toEqual(422);
+        expect(res.body.message).toBe(message);
+      });
+
+
+      it('should return 422 if validation fails', async () => {
```
Comment (johnmutuma5) [inline src/modules/requests/__tests__/requests.test.js]: @Spectrumsun, my first impression was that this is one test. I suggest you make the description in `it` more specific to each test. Also the `newRequest` object seems very similar across all tests; I think it's proper to define it within a `beforeEach` or `beforeAll`, whichever suites you best, in a `describe` block for these tests and then just adjust the changing property(s) for a particular test to reduce their bulkiness: e.g. in the second test above, 
```
it('blablabah', () => {
  ...
  newRequest.arrivalDate = '2018-09-16';
  newRequest.oneWayTrip = true;
  ...
}
 ```

File: src/modules/requests/__tests__/requests.test.js [hunk 3]
```diff
@@ -417,8 +423,137 @@ describe('Requests Controller', () => {
...
+        const message = 'Trip type must be a return , oneWay or multi Trip';
+        const res = await request(app)
+          .post('/api/v1/requests')
+          .set('authorization', token)
+          .send({ ...newRequest });
+        expect(res.status).toEqual(422);
+        expect(res.body.message).toBe(message);
+      });
+
+      it('should return 422 if Arrival Date is not added for a return trip', async () => {
```
Comment (johnmutuma5) [inline src/modules/requests/__tests__/requests.test.js]: @Spectrumsun , the newRequest object seems very similar across the six test cases; I think it's proper to define it within a `beforeEach` in a describe block for these tests and then just adjust the changing property(s) for a particular test case to reduce their bulkiness: e.g. in this test,
``` js
it('...if Arrival...not added...return trip', async () => {
  newRequest.tripType ='return';
  delete newRequest.arrivalDate;
  // ... 
}
```

File: src/modules/requests/__tests__/requests.test.js [hunk 4]
```diff
@@ -417,8 +423,137 @@ describe('Requests Controller', () => {
...
+          tripType: 'return',
+          createdAt: '2018-08-16 012:11:52.181+01',
+          updatedAt: '2018-08-16 012:11:52.181+01'
+        };
+        const message = 'Arrival Date must be a date';
+        const res = await request(app)
+          .post('/api/v1/requests')
+          .set('authorization', token)
+          .send({ ...newRequest });
+        console.log(res.body);
```
Comment (johnmutuma5) [inline src/modules/requests/__tests__/requests.test.js]: @Spectrumsun, you also forgot this one here too.

Comment (Spectrumsun) [reply src/modules/requests/__tests__/requests.test.js]: fixed

File: src/modules/requests/__tests__/requests.test.js (hunk 3 above)
Comment (Spectrumsun) [reply src/modules/requests/__tests__/requests.test.js]: the content is different from what I can see and am not the one who wrote does test cases. they might have been a reason why they wrote it like that

File: src/modules/requests/__tests__/requests.test.js (hunk 4 above)
Comment (albert85) [reply src/modules/requests/__tests__/requests.test.js]: @Spectrumsun when i tried to run `yarn db:rollback` this is the error i got.

screen shot 2018-09-12 at 4 56 45 pm

Comment (albert85) [reply src/modules/requests/__tests__/requests.test.js]: Don't worry @Spectrumsun it is from my end

File: src/modules/requests/__tests__/requests.test.js (hunk 3 above)
Comment (johnmutuma5) [reply src/modules/requests/__tests__/requests.test.js]: @Spectrumsun, if that's so then let's not mind that. Good job on the feature. Keep it up.

**Comment by samuelkubai** [comment]
@Spectrumsun could you please resolve the conflicts.

**Comment by Spectrumsun** [comment]
@samuelkubai  fixed

**Comment by samuelkubai** [comment]
@Spectrumsun this is being implemented by @amarachukwu-agbo's PR.

Closing this PR.

**Pull request closed without merging**